File: reboot/examples/bank-nodejs/backend/src/bank_servicer.ts
import { TransactionContext, allow } from "@reboot-dev/reboot";
import { Account } from "../../api/bank/v1/account_rbt.js";
import { Bank } from "../../api/bank/v1/bank_rbt.js";

function randomIntFromInterval(min: number, max: number): number {
  // Min / max inclusive.
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export class BankServicer extends Bank.Servicer {
  authorizer() {
    return allow();
  }

  async PickNewAccountId(): Promise<string> {
    // Transactions normally observe state through Reader calls. However for
    // convenience, it is possible to do an inline read of the state of this
    // state machine, which is like calling a Reader that simply returns the
    // whole state of the state machine.

    while (true) {
      const newAccountId = String(randomIntFromInterval(1000000, 9999999));
      if (!this.state.accountIds.includes(newAccountId)) {
        return newAccountId;
      }
    }
  }

  async signUp(
    context: TransactionContext,
    request: Bank.SignUpRequest
  ): Promise<Bank.PartialSignUpResponse> {
    const newAccountId = await this.PickNewAccountId();

    // Transactions like writers can alter state directly.
    this.state.accountIds.push(newAccountId);

    // Let's go create the account.
    const [account, response] = await Account.open(context, newAccountId, {
      customerName: request.customerName,
    });

    return { accountId: newAccountId };
  }

  async transfer(
    context: TransactionContext,
    request: Bank.TransferRequest
  ): Promise<Bank.PartialTransferResponse> {
    const fromAccount = Account.ref(request.fromAccountId);
    const toAccount = Account.ref(request.toAccountId);

    await fromAccount.withdraw(context, { amount: request.amount });
    await toAccount.deposit(context, { amount: request.amount });

    return {};
  }
}

File: reboot/examples/monorepo/bank/backend/src/bank_servicer.py
import logging
import uuid
from bank.v1.account_rbt import Account
from bank.v1.bank_rbt import (
    Bank,
    SignUpRequest,
    SignUpResponse,
    TransferRequest,
    TransferResponse,
)
from reboot.aio.auth.authorizers import allow
from reboot.aio.contexts import TransactionContext

logging.basicConfig(level=logging.INFO)


class BankServicer(Bank.Servicer):

    def authorizer(self):
        return allow()

    async def sign_up(
        self,
        context: TransactionContext,
        request: SignUpRequest,
    ) -> SignUpResponse:
        # Generating an account ID so that we can demonstrate setting
        # the account ID explicitly. Alternatively you can just call
        # `construct()` without any args and Reboot will generate a
        # unique ID for you.
        new_account_id = str(uuid.uuid4())

        # Let's go create the account.
        account, _ = await Account.open(
            context,
            new_account_id,
            customer_name=request.customer_name,
        )

        # Transactions like writers can alter state directly.
        self.state.account_ids.append(account.state_id)

        return SignUpResponse(account_id=account.state_id)

    async def transfer(
        self,
        context: TransactionContext,
        request: TransferRequest,
    ) -> TransferResponse:
        from_account = Account.ref(request.from_account_id)
        to_account = Account.ref(request.to_account_id)

        await from_account.withdraw(context, amount=request.amount)
        await to_account.deposit(context, amount=request.amount)

        return TransferResponse()
